### PR TITLE
Make path finding index-based

### PR DESCRIPTION
**Pull request by glebm**
Reduces the size of the `PathNodes` array from 28 KiB to just 8 KiB.

Also reduces the size of `pnode_tblptr` from `300 * sizeof(void *)` to `300 * 2` bytes.

### CODE REVIEW

File: Source/path.cpp [hunk 1]
```diff
@@ -11,116 +11,143 @@
 #include "objects.h"
 
 namespace devilution {
-
 namespace {
 
+constexpr size_t MaxPathNodes = 300;
+
+struct PathNode {
+	static constexpr uint16_t InvalidIndex = std::numeric_limits<uint16_t>::max();
```
Comment (ephphatha) [inline Source/path.cpp:19]: Is it worth using `uint8_t` as an index type here? values currently will always be in the range 0-24

```suggestion
	using IndexType = uint_fast8_t
	static constexpr IndexType InvalidIndex = std::numeric_limits<IndexType>::max()
```

File: Source/path.cpp [hunk 2]
```diff
@@ -11,116 +11,143 @@
...
+
+	void addChild(uint16_t childIndex)
+	{
+		size_t index = 0;
+		for (; index < 8; ++index) {
+			if (childIndices[index] == InvalidIndex)
+				break;
+		}
+		childIndices[index] = childIndex;
+	}
```
Comment (ephphatha) [inline Source/path.cpp:43]: While it's impossible to hit in the current code as tiles only have 8 neighbours (so this is called a max of 8 times for a given node) this tries to write past the end of the array if all child nodes are full. Would it be worth restructuring the function to set and early return in the loop? e.g.
```suggestion
	void addChild(uint16_t childIndex)
	{
		for (auto &index : childIndices) {
			if (index == InvalidIndex) {
				index = childIndex;
				return;
			}
		}
	}
```

File: Source/path.cpp (hunk 1 above)
Comment (glebm) [reply Source/path.cpp:19]: `MaxPathNodes` is 300, are you saying we can reduce `MaxPathNodes` to 24?

File: Source/path.cpp (hunk 2 above)
Comment (glebm) [reply Source/path.cpp:43]: We can detect the writes past-the-end with ASAN whereas that version would just do nothing silently. Should I add an assert?

File: Source/path.cpp (hunk 1 above)
Comment (ephphatha) [reply Source/path.cpp:19]: My mistake, I missed the define and use of MaxPathNodes😂

File: Source/path.cpp (hunk 2 above)
Comment (ephphatha) [reply Source/path.cpp:43]: an assert sounds good

Comment (glebm) [reply Source/path.cpp]: Done

**Review by ephphatha — APPROVED** [review]

File: Source/platform/locale.cpp [hunk 3]
```diff
@@ -11,6 +11,9 @@
 #elif defined(__3DS__)
 #include "platform/ctr/locale.hpp"
 #elif defined(_WIN32)
+// Suppress definitions of `min` and `max` macros by <windows.h>:
+#define NOMINMAX 1
```
Comment (AJenbo) [inline Source/platform/locale.cpp:15]: This appears to already be there

Comment (glebm) [reply Source/platform/locale.cpp:15]: Rebase accident, thanks @ephphatha for fixing!

**Pull request merged**